src/rust/engine/graph/src/lib.rs

```diff
@@ -653,21 +653,21 @@ impl<N: Node> Graph<N> {
           // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
           // without a much more complicated algorithm.
           let potential_dst_id = inner.ensure_entry(dst_node.clone());
-          match Self::detect_cycle(src_id, potential_dst_id, &mut inner) {
-            Ok(true) => {
-              // Cyclic dependency: declare a dependency on a copy of the Node that is marked Cyclic.
-              inner.ensure_entry(EntryKey::Cyclic(dst_node))
-            }
-            Ok(false) => {
-              // Valid dependency.
-              trace!(
-                "Adding dependency from {:?} to {:?}",
-                inner.entry_for_id(src_id).unwrap().node(),
-                inner.entry_for_id(potential_dst_id).unwrap().node()
-              );
-              potential_dst_id
-            }
-            Err(err) => return futures::future::err(err).to_boxed(),
+          if let Some(cycle_path) = Self::detect_cycle(src_id, potential_dst_id, &mut inner) {
```

Comment:
Maybe we could extract the logic from detecting a cycle to forming the error message (with the cycle represented as a string) into its own function.

That would allow to create a simple unit test to represent an example like the one @wisechengyi quoted below.

It would help visualise what we're outputting to the user, and possibly refine it (for instance, by doing something more subtle than `e.node().to_string()` if appropriate)